### PR TITLE
device map legacy attention block weight conversion

### PR DESCRIPTION
**Pull request by williamberman**
re: https://github.com/huggingface/diffusers/issues/3740

this is not ideal but afaik this is the only way to solve this without exposing functionality in accelerate

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def __init__(
         self.upcast_softmax = upcast_softmax
         self.rescale_output_factor = rescale_output_factor
         self.residual_connection = residual_connection
+        self.dropout = dropout
```
Comment (williamberman) [inline src/diffusers/models/attention_processor.py:81]: Added so can re-create the dropout module when converting back to new weight format

File: src/diffusers/models/modeling_utils.py [hunk 2]
```diff
@@ -646,15 +646,47 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+                            dtype=torch_dtype,
+                        )
+                    except AttributeError as e:
+                        # When using accelerate loading, we do not have the ability to load the state
+                        # dict and rename the weight names manually. Additionally, accelerate skips
+                        # torch loading conventions and directly writes into `module.{_buffers, _parameters}`
+                        # (which look like they should be private variables?), so we can't use the standard hooks
+                        # to rename parameters on load. We need to mimic the original weight names so the correct
+                        # attributes are available. After we have loaded the weights, we convert the deprecated
+                        # names to the new non-deprecated names. Then we _greatly encourage_ the user to convert
```
Comment (sayakpaul) [inline src/diffusers/models/modeling_utils.py:666]: Do we have guidance available for the users on how they should perform the conversion?

Comment (sayakpaul) [reply src/diffusers/models/modeling_utils.py:666]: Ah nevermind. I guess you meant once we load the old attention block weight names and run conversion internally, and suggest users save the pipeline. Right?

File: tests/models/test_attention_processor.py [hunk 3]
```diff
@@ -73,3 +76,44 @@ def test_only_cross_attention(self):
...
 
         self.assertTrue((only_cross_attn_out != self_and_cross_attn_out).all())
+
+
+class DeprecatedAttentionBlockTests(unittest.TestCase):
+    def test_conversion_when_using_device_map(self):
+        pipe = DiffusionPipeline.from_pretrained("hf-internal-testing/tiny-stable-diffusion-pipe", safety_checker=None)
+
+        pre_conversion = pipe(
+            "foo",
```
Comment (sayakpaul) [inline tests/models/test_attention_processor.py:86]: Killer prompt.

**Review by sayakpaul — APPROVED** [review]
Concrete! Nice tests.

File: src/diffusers/models/modeling_utils.py (hunk 2 above)
Comment (williamberman) [reply src/diffusers/models/modeling_utils.py:666]: Yeah just this blurb here
```
  f"Taking `{str(e)}` while using `accelerate.load_checkpoint_and_dispatch` to mean {pretrained_model_name_or_path}"
                                " was saved with deprecated attention block weight names. We will load it with the deprecated attention block"
                                " names and convert them on the fly to the new attention block format. Please re-save the model after this conversion,"
                                " so we don't have to do the on the fly renaming in the future. If the model is from a hub checkpoint,"
                                " please also re-upload it or open a PR on the original repository."
```

File: src/diffusers/models/modeling_utils.py [hunk 4]
```diff
@@ -646,15 +646,47 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+                        # the weights so we don't have to do this again.
+
+                        if "'Attention' object has no attribute" in str(e):
+                            logger.warn(
+                                f"Taking `{str(e)}` while using `accelerate.load_checkpoint_and_dispatch` to mean {pretrained_model_name_or_path}"
+                                " was saved with deprecated attention block weight names. We will load it with the deprecated attention block"
+                                " names and convert them on the fly to the new attention block format. Please re-save the model after this conversion,"
+                                " so we don't have to do the on the fly renaming in the future. If the model is from a hub checkpoint,"
+                                " please also re-upload it or open a PR on the original repository."
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:675]:
```suggestion
                                " please also re-upload both the model file and config or open a PR on the original repository: {pretrained_model_name_or_path}."
```

File: src/diffusers/models/modeling_utils.py [hunk 5]
```diff
@@ -889,3 +921,53 @@ def recursive_find_attn_block(name, module):
                 state_dict[f"{path}.to_out.0.weight"] = state_dict.pop(f"{path}.proj_attn.weight")
             if f"{path}.proj_attn.bias" in state_dict:
                 state_dict[f"{path}.to_out.0.bias"] = state_dict.pop(f"{path}.proj_attn.bias")
+
+    def _temp_convert_self_to_deprecated_attention_blocks(self):
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:925]:
```suggestion
    def _temp_convert_self_to_deprecated_attention_blocks(self):
        " This method should only be used to convert deprecated checkpoints and will be removed in a future version."
```

File: src/diffusers/models/modeling_utils.py [hunk 6]
```diff
@@ -889,3 +921,53 @@ def recursive_find_attn_block(name, module):
...
+            # We don't _have_ to delete the old attributes, but it's helpful to ensure
+            # that _all_ the weights are loaded into the new attributes and we're not
+            # making an incorrect assumption that this model should be converted when
+            # it really shouldn't be.
+            del module.to_q
+            del module.to_k
+            del module.to_v
+            del module.to_out
+
+    def _undo_temp_convert_self_to_deprecated_attention_blocks(self):
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:952]:
```suggestion
    def _undo_temp_convert_self_to_deprecated_attention_blocks(self):
        " This method should only be used to convert deprecated checkpoints and will be removed in a future version of diffusers."
```

File: src/diffusers/models/modeling_utils.py [hunk 7]
```diff
@@ -646,15 +646,47 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+                        # When using accelerate loading, we do not have the ability to load the state
+                        # dict and rename the weight names manually. Additionally, accelerate skips
+                        # torch loading conventions and directly writes into `module.{_buffers, _parameters}`
+                        # (which look like they should be private variables?), so we can't use the standard hooks
+                        # to rename parameters on load. We need to mimic the original weight names so the correct
+                        # attributes are available. After we have loaded the weights, we convert the deprecated
+                        # names to the new non-deprecated names. Then we _greatly encourage_ the user to convert
+                        # the weights so we don't have to do this again.
+
+                        if "'Attention' object has no attribute" in str(e):
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:669]: pretty hacky, but OK! Let's leave it for now :-)

Comment (williamberman) [reply src/diffusers/models/modeling_utils.py:669]: yeah I cringed while writing this 🙃
